File: Model/model/spaces.py
from .types import SecondsType, XType, NumberIterationsType, ObjectiveValueType, ActionNameType, ObjectiveDerivativeValueType
from typing import TypedDict

TerminatingSpace = TypedDict('Terminating Space', {})
EmptySpace = TypedDict('Empty Space', {})
InitializationSpace = TypedDict('Initialization Space', {})
UpdateXSpace = TypedDict('Update X Space', {'x': XType})
UpdateYSpace = TypedDict('Update Y Space', {'y': ObjectiveValueType})
UpdateYPrimeSpace = TypedDict('Update Y Prime Space', {'y_prime': ObjectiveDerivativeValueType})
NoneSpace = TypedDict('None Space', {})
ComputationTimeMetricSpace = TypedDict('Computation Time Metric Space', {'simulation_time': SecondsType, 'computation_time': SecondsType, 'action_name': ActionNameType})
StateMetricSpace = TypedDict('State Metric Space', {'simulation_time': SecondsType, 'x': XType, 'y': ObjectiveValueType, 'y_prime': ObjectiveDerivativeValueType, 'iteration_step': NumberIterationsType})

File: Julia Model/spaces.jl
include("../../BlockScience/cadCAD.jl/src/spaces.jl")
include("types.jl")
using .Spaces: generate_space_type

generate_space_type((x=XType,), "Update X Space")
generate_space_type((y=ObjectiveValueType,), "Update Y Space")
generate_space_type((y_prime=ObjectiveDerivativeValueType,), "Update Y Prime Space")
generate_space_type((simulation_time=SecondsType, computation_time=SecondsType, action_name=ActionNameType,), "Computation Time Metric Space")
generate_space_type((simulation_time=SecondsType, x=XType, y=ObjectiveValueType, y_prime=ObjectiveDerivativeValueType, iteration_step=NumberIterationsType,), "State Metric Space")
